book/html.md

```diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
+
+[^ignore-them]: Real browsers retain whitespace nodes: whitespace is
+    significant inside `<pre>` tags or in cases like the difference
+    between `make<span>up</span>` and `make <span>up</span>`. Our
+    browser won't support that, and ignoring all whitespace tags
+    simplifies [later chapters](layout.md) by avoiding a special-case
+    for whitespace-only text tags.
 
-Why aren't these open elements closed? Well, most of them (like
-`<meta>` and `<link>`) are what are called self-closing: you don't
-ever write `</meta>` or `</link>`. These tags don't need a close tag
-because they never surround content. Let's add that to our parser:
+``` {.python indent=4}
+def add_text(self, text):
+    if text.isspace(): return
+    # ...
+```
 
+The parsed HTML tree now looks like this:
 
-``` {.python indent=8 replace=parent)/parent%2c%20tok.attributes)}
-# ...
-elif tok.tag in SELF_CLOSING_TAGS:
-    node = ElementNode(tok.tag, parent)
-    parent.children.append(node)
+``` {.example}
+<html lang="en-US" xml:lang="en-US" >
+   <head >
+     <meta charset="utf-8" / >
+       <meta name="generator" content="pandoc" / >
+         <meta name="viewport" content="width=device-width,initial-scale=1.0,user-scalable=yes" / >
+           <meta name="author" content="Pavel Panchekha &amp; Chris Harrelson" / >
+             <link rel="stylesheet" href="book.css" / >
+               <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Vollkorn%7CLora&display=swap" / >
+                 <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Vollkorn:400i%7CLora:400i&display=swap" / >
+                   <title >
 ```
 
-Use the following `SELF_CLOSING_TAGS` list, straight from the
-[standard][html5-void-elements]:[^void-elements]
+Why's everything so deeply indented? Why aren't these open elements ever closed?
+
+::: {.further}
+SGML document type declarations had a URL to define the valid tags.
+Browsers use the absense of a document type declaration to
+[identify][quirks-mode] very old, pre-SGML versions of
+HTML,[^almost-standards-mode] but don't need the URL, so `<!doctype
+html>` is the best document type declaration today.
+:::
+
+[quirks-mode]: https://developer.mozilla.org/en-US/docs/Web/HTML/Quirks_Mode_and_Standards_Mode
+
+[^almost-standards-mode]: There's also a crazy thing called "[almost
+    standards][limited-quirks]" or "limited quirks" mode, due to a
+    backwards-incompatible change in table cell vertical layout. Yes.
+    I don't need to make these up!
+
+[limited-quirks]: https://hsivonen.fi/doctype/
+
+Self-closing tags
+=================
+
+Elements like `<meta>` and `<link>` are what are called self-closing: you don't
+ever write `</meta>` or `</link>`, because these tags don't surround content. To get a reasonable-looking tree out of our parser, we'll need special support for them. In HTML, there's a [fixed list][html-void-elements] of these self-closing tags:[^void-elements]
```

Comment:
```suggestion
ever write `</meta>` or `</link>`, because these tags don't surround content. To get a reasonable-looking tree out of our parser, we'll need special support for them. In HTML, there's a [fixed list][html5-void-elements] of these self-closing tags:[^void-elements]
```